llvm/lib/Target/AMDGPU/SIInsertWaitcnts.cpp

```diff
@@ -1824,7 +1824,9 @@ bool SIInsertWaitcnts::generateWaitcntInstBefore(MachineInstr &MI,
   // Verify that the wait is actually needed.
   ScoreBrackets.simplifyWaitcnt(Wait);
 
-  if (ForceEmitZeroFlag)
+  // When forcing emit, we need to skip terminators of a MBB because that would
+  // break the terminators of the MBB.
+  if (ForceEmitZeroFlag && !MI.isTerminator())
```

Comment:
We should be able to assume all terminators are not memory instructions

Reply:
which means we don't actually need a wait even before the first terminator?